ADAL/src/validation/ADAuthorityValidation.m

```diff
@@ -198,59 +180,124 @@ - (void)validateAuthority:(ADRequestParameters*)requestParams
+        // scenarios.
+        if (dispatch_semaphore_wait(dsem, DISPATCH_TIME_NOW) != 0)
         {
-            [self addValidAuthority:authority];
+            // Only bother logging if we have to wait on the queue.
+            AD_LOG_INFO(@"Waiting on Authority Validation Queue", requestParams.correlationId, nil);
+            dispatch_semaphore_wait(dsem, DISPATCH_TIME_FOREVER);
+            AD_LOG_INFO(@"Returned from Authority Validation Queue", requestParams.correlationId, nil);
         }
-        
-        completionBlock(verified, error);
-    }];
+    });
 }
 
+- (void)requestAADValidation:(NSURL *)authority
+               requestParams:(ADRequestParameters *)requestParams
+             completionBlock:(ADAuthorityValidationCallback)completionBlock
+{
+    // Before we make the request, check the cache again, as these requests happen on a serial queue
+    // and it's possible we were waiting on a request that got the information we're looking for.
+    ADAadAuthorityCacheRecord *record = [_aadCache checkCache:authority];
+    if (record)
+    {
+        completionBlock(record.validated, record.error);
+        return;
+    }
+    
+    NSString *trustedHost = s_kTrustedAuthorityWorldWide;
+    NSString *authorityHost = authority.adHostWithPortIfNecessary;
+    if ([_whitelistedAADHosts containsObject:authorityHost])
+    {
+        trustedHost = authorityHost;
+    }
+    
+    [ADAuthorityValidationRequest requestMetadataWithAuthority:authority.absoluteString
+                                                   trustedHost:trustedHost
+                                                       context:requestParams
+                                               completionBlock:^(NSDictionary *response, ADAuthenticationError *error)
+     {
+         if (error)
+         {
+             completionBlock(NO, error);
+             return;
+         }
+         
+         NSString *oauthError = response[@"error"];
+         if (oauthError)
+         {
+             ADAuthenticationError *adError =
+             [ADAuthenticationError errorFromAuthenticationError:AD_ERROR_DEVELOPER_AUTHORITY_VALIDATION
+                                                    protocolCode:oauthError
+                                                    errorDetails:response[@"error_details"]
+                                                   correlationId:requestParams.correlationId];
+             
+             // If the error is something other than invalid_instance then something wrong is happening
+             // on the server.
+             if ([oauthError isEqualToString:@"invalid_instance"])
+             {
+                 [_aadCache addInvalidRecord:authority oauthError:adError context:requestParams];
+             }
+             
+             completionBlock(NO, adError);
+             return;
+         }
+         
+         [_aadCache processMetadata:response[@"metadata"] authority:authority context:requestParams];
```

Comment:
I'd prefer checking metadata type somewhere, because in processMetaData first thing that's done is "if (metadata.count == 0)", which might crash unless metadata is NSArray type (imagine it comes as NSString or NSNull)....